### PR TITLE
NET: Yahya/5147-deduplicate-channels-part-2

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This PR is the second part of deduplicating channels involved in the network topology and makes it feature-complete. The topology of the networking layer constitutes overlapping channels, where nodes are able to subscribe to a channel and disseminate messages within that channel. Fanout of a node (i.e., number of outgoing connections) is correlated with the number of topics the node subscribes.

Channels only differ in the set of roles involved in them. A node subscribes to a set of channels based on its role. We identified that there are several duplicate channels in terms of the set of involved roles. So, this PR deduplicates such channel, so that topology is constructed only for unique channels based on their set of involved roles.

### CODE REVIEW

File: network/topology/randomizedTopology.go [hunk 1]
```diff
@@ -67,8 +67,9 @@ func NewRandomizedTopology(nodeID flow.Identifier,
 // connected graph of nodes that enables them talking to each other. This should be done with a very high probability
 // in randomized topology.
 func (r RandomizedTopology) GenerateFanout(ids flow.IdentityList) (flow.IdentityList, error) {
-	myChannels := r.subMngr.Channels()
 	if len(myChannels) == 0 {
+	myUniqueChannels := engine.UniqueChannels(r.subMngr.Channels())
```
Comment (yhassanzadeh13) [inline network/topology/randomizedTopology.go:71]: deduplicates channels the node subscribed to.

File: engine/channels_test.go [hunk 2]
```diff
@@ -87,3 +87,40 @@ func TestIsClusterChannel(t *testing.T) {
...
+// UniqueChannels are unique based on their set of involved roles.
+// We use the identifier of RoleList to determine their uniqueness.
+func TestUniqueChannels_Uniqueness(t *testing.T) {
+	for _, role := range flow.Roles() {
+		uniques := UniqueChannels(ChannelsByRole(role))
+
+		visited := make(map[flow.Identifier]struct{})
+		for _, channel := range uniques {
+
+			if _, ok := IsClusterChannel(channel); !ok {
```
Comment (vishalchangrani) [inline engine/channels_test.go:101]:
```suggestion
			if _, ok := IsClusterChannel(channel); ok {
			continue //only considering non-cluster channel in this test case
			}
```

File: engine/channels.go [hunk 3]
```diff
@@ -49,6 +49,36 @@ func ChannelsByRole(role flow.Role) network.ChannelList {
 	return channels
 }
 
+// UniqueChannels returns list of non-cluster channels with a unique RoleList accompanied
+// with the list of all cluster channels.
```
Comment (vishalchangrani) [inline engine/channels.go:53]:
```suggestion
// with the list of all cluster channels.
// e.g. is channel X and Y both are non-cluster channels and have role IDs [A,B,C] then only one of them will be in the returned list.
```

**Review by vishalchangrani — APPROVED** [review]
lgtm

File: engine/channels.go [hunk 4]
```diff
@@ -148,10 +179,10 @@ func initializeChannelRoleMap() {
 	channelRoleMap[consensusClusterPrefix] = flow.RoleList{flow.RoleCollection}
 }
 
-// IsClusterChannel returns true if channel is cluster-based.
+// ClusterChannel returns true if channel is cluster-based.
```
Comment (vishalchangrani) [inline engine/channels.go:182]: also returns the channel prefix (not just true or false)

Comment (yhassanzadeh13) [reply engine/channels.go:182]: right, but it already has this explanation in the subsequent lines:
```
// ClusterChannel returns true if channel is cluster-based.
// At the current implementation, only collection nodes are involved in a cluster-based channels.
// If the channel is a cluster-based one, this method also strips off the channel prefix and returns it.
```

File: engine/channels.go [hunk 5]
```diff
@@ -49,6 +49,37 @@ func ChannelsByRole(role flow.Role) network.ChannelList {
 	return channels
 }
 
+// UniqueChannels returns list of non-cluster channels with a unique RoleList accompanied
+// with the list of all cluster channels.
+// e.g. is channel X and Y both are non-cluster channels and have role IDs [A,B,C] then only one of them will be in the returned list.
```
Comment (jordanschalm) [inline engine/channels.go:54]:
```suggestion
// e.g. if channel X and Y both are non-cluster channels and have role IDs [A,B,C] then only one of them will be in the returned list.
```

File: engine/channels_test.go [hunk 6]
```diff
@@ -67,23 +67,62 @@ func TestGetChannelByRole(t *testing.T) {
...
+			require.False(t, duplicate)
+
+			visited[id] = struct{}{}
+		}
+	}
+}
+
+// TestUniqueChannels_ClusterChannels verifies that if cluster channels have the RoleList the same as
+// single non-cluster channel, then all cluster channels as well as the one non-cluster channel are returned
+// by the UniqueChannels. In other words, neither cluster channels nor non-cluster ones are not deduplicated in the
```
Comment (jordanschalm) [inline engine/channels_test.go:117]:
```suggestion
// by the UniqueChannels. In other words, neither cluster channels nor non-cluster ones are de-duplicated in the
```

**Review by jordanschalm — APPROVED** [review]

**Pull request merged**